include/envoy/ssl/context_config.h

```diff
@@ -53,6 +55,11 @@ class ContextConfig {
    */
   virtual const std::string& privateKeyFile() const PURE;
 
+  /**
+   * @return The CRL to check if a cert is revoked.
+   */
+  virtual const std::vector<bssl::UniquePtr<X509_CRL>>& certificateRevocationLists() const PURE;
```

Comment:
This is (common?) interface and not implementation, so I don't think you should use BoringSSL-specific types here.

Reply:
So, there was a request [in the previous PR](https://github.com/envoyproxy/envoy/pull/2077#discussion_r151738924) to have the class hold a `bssl::UniquePtr<X509_CRL>` so that invalid data would be caught sooner.  Agreed that it's a bit ugly to have BoringSSL-specific types in the interface, so happy to make whichever changes you / @ggreenway would prefer.

Reply:
I agree with @PiotrSikora that we shouldn't expose boringssl at this interface level. I haven't fully reviewed the change but can we find a way to hide this? We might need to add a CRL interface also.

Reply:
So, here's three possible options:

1. Have `ContextConfig` contain a `DataSource`, or just the raw bytes for a CRL.
  **Pros:** we don't expose BoringSSL in this interface; simple.
  **Cons:** if the CRL is invalid, we won't find out until we try to instantiate a listener (since that's when we turn the bytes into a CRL).

2. Have `ContextConfig` contain the BoringSSL types.
  **Pros:** we parse the CRL ~immediately and thus can error at config load time; already exists.
  **Cons:** doesn't feel like `ContextConfig` should have BoringSSL types.

3. Introduce a `CRL` interface (or something similar), parse the CRL up-front, wrap into the interface and use that here.
  **Pros:** probably can get the best of both worlds (verifying the CRL up-front and hiding the concrete type).
  **Cons:** complicated; need to make sure that whatever parses the config does so into a format that the `ContextConfig` implementation understands (i.e. there's coupling between config loading and the `ContextConfigImpl`).

Thoughts?

Reply:
Yeah agreed this is a little tricky. Some random ideas:

1) Variant of (1) where we expose the raw bytes, but also verify that it's valid inside the config and throw if not. (Basically we would end up loading the CRL twice, but I think that's probably not a big deal).

2) Some type of visitor pattern, where the `ContextImpl` calls back into the config and the config sets up the SSL context. This would require some dynamic_cast to get at the `SSL` inside the config impl, but we are already doing that in various places inside config impl for similar reasons, so I think it's probably OK.

Reply:
FWIW, https://github.com/envoyproxy/envoy/pull/2248 is just loading and exposing raw bytes, without verifying the content. We could start loading them twice, but this will start being costly at scale, and will hurt loading times.

Reply:
I think the most pragmatic solution is to just expose raw bytes in this interface, and catch any data-validation errors in the ContextImpl.  We already catch many similar types of errors there.